### PR TITLE
[JENKINS-47609] Add clone option "core.longpaths" to enable long file names.

### PR DESCRIPTION
**Pull request by rishabhBudhouliya**
## [JENKINS-47609](https://issues.jenkins-ci.org/browse/JENKINS-47609) 

Please refer to this PR: https://github.com/jenkinsci/git-plugin/pull/856  

This PR includes the changes in CloneCommand, CliGitAPIImpl and JGitAPIImpl to implement a new clone behavior which enables "core.longpaths".

### CODE REVIEW

**Comment by MarkEWaite** [comment]
You'll need to resolve the conflicts or rebase the changes on the tip of the upstream master branch.

**Comment by MarkEWaite** [comment]
You may want to consider installing the [`hub`](https://github.com/github/hub/blob/master/README.md) command in your local development environment.  It simplifies interactions with GitHub.  You update from the upstream repository with a single command:

```
$ hub sync
```

You create a new pull request with a single command:

```
$ hub pull-request
```

You evaluate someone else's pull request with a single command:

```
$ hub pr checkout 523
```

**Comment by rishabhBudhouliya** [comment]
I was actually trying to solve this with hub, I think I've done something wrong.

File: src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java [hunk 1]
```diff
@@ -514,6 +514,132 @@ public void test_setCommitter() throws Exception {
...
+        check_remote_url("upstream");
+        assertBranchesExist(w.git.getBranches(), "master");
+        final String alternates = ".git" + File.separator + "objects" + File.separator + "info" + File.separator + "alternates";
+        assertFalse("Alternates file found: " + alternates, w.exists(alternates));
+    }
+
+    public void test_clone_longPath() throws IOException, InterruptedException
+    {
+        w.git.clone_().url(localMirror()).repositoryName("upstream").longPath(true).execute();
+        assertThat("true", is(getConfigValue(w.repo, "core.longpaths")));
```
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java]: Hamcrest assertion message will display the right hand value as the expected value and the left hand value as the actual value. 
```suggestion
        assertThat(getConfigValue(w.repo, "core.longpaths"), is("true"));
```

File: src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java [hunk 2]
```diff
@@ -514,6 +514,132 @@ public void test_setCommitter() throws Exception {
...
+        createRevParseBranch(); // Verify JENKINS-32258 is fixed
+        w.git.checkout().ref("origin/master").branch("master").execute();
+        check_remote_url("origin");
+        assertBranchesExist(w.git.getBranches(), "master");
+        final String alternates = ".git" + File.separator + "objects" + File.separator + "info" + File.separator + "alternates";
+        assertFalse("Alternates file found: " + alternates, w.exists(alternates));
+        assertFalse("Unexpected shallow clone", w.cgit().isShallowRepository());
+    }
+
+    public void test_checkout_exception() throws Exception {
```
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java]: This test is unrelated to longpaths and should be removed.

File: src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java [hunk 3]
```diff
@@ -514,6 +514,132 @@ public void test_setCommitter() throws Exception {
...
+     *   repositoryName(String) - if omitted, CliGit does not set a remote repo name
+     *   shallow() - no relevant assertion of success or failure of this argument
+     *   shared() - not implemented on CliGit, not verified on JGit
+     *   reference() - implemented on JGit, not verified on either JGit or CliGit
+     *
+     * CliGit and JGit both require the w.git.checkout() call
+     * otherwise no branch is checked out. That is different than the
+     * command line git program, but consistent within the git API.
+     */
+    public void test_clone() throws Exception
```
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java]: This test is unrelated to long paths and should be removed

File: src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java [hunk 4]
```diff
@@ -514,6 +514,132 @@ public void test_setCommitter() throws Exception {
...
+        final String SHA1 = "feedbeefbeaded";
+        try {
+            w.git.checkout().ref(SHA1).branch("master").execute();
+            fail("Expected checkout exception not thrown");
+        } catch (GitException ge) {
+            assertEquals("Could not checkout master with start point " + SHA1, ge.getMessage());
+        }
+    }
+
+    public void test_clone_repositoryName() throws IOException, InterruptedException
```
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java]: This test is unrelated to long paths and should be removed.

File: src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java [hunk 5]
```diff
@@ -514,6 +514,132 @@ public void test_setCommitter() throws Exception {
...
+        assertFalse("Alternates file found: " + alternates, w.exists(alternates));
+    }
+
+    public void test_clone_longPath() throws IOException, InterruptedException
+    {
+        w.git.clone_().url(localMirror()).repositoryName("upstream").longPath(true).execute();
+        assertThat("true", is(getConfigValue(w.repo, "core.longpaths")));
+    }
+
+    public void test_clone_shallow() throws Exception
```
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java]: Unrelated to long paths, remove this block

File: src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java (hunk 1 above)
Comment (MarkEWaite) [reply src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java]: This test should move to one of the GitClient*Test classes so that it tests all the implementations and is expressed as JUnit 4 test.  Then you can rest GitAPITestCase to match the contents of the master branch.

Comment (MarkEWaite) [reply src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java]: I would also like a test that performs a clone into a directory with an absolute path longer than 260 characters.  The test for the config value is good and necessary, but a test that things are well behaved is even better.

**Comment by rishabhBudhouliya** [comment]
@MarkEWaite thanks for the review, I have updated the PR with the relevant changes (I'll try to make the code cleaner and semantically better next time).

**Comment by rishabhBudhouliya** [comment]
The new commit: [add3695](https://github.com/jenkinsci/git-client-plugin/pull/523/commits/add3695a59e3c7c38fdc8af9f32012344a3a2934) address the issue of workspace directory not being able to be renamed to a 260 char file path. (Build was potentially failing because of this issue)

**Comment by rishabhBudhouliya** [comment]
@MarkEWaite I am facing an issue on Windows platform, 
```java.io.IOException: Cannot run program "git" (in directory "C:\Jenkins\workspace\Plugins_git-client-plugin_PR-523\target\tmp\tehdfmufzlrimdyxmzetepzxegbupzxewqtnlhwiwtcugemiizqsstahyyqufwpdjyolbkegykuhbwdgxmvezycngdpribpzvsiltzsdinfbokkwdnhsmfeeqcioeulxkqeagtjhqzuukqfglubmmcrbjhgifowrlbbxvbgdlqmuzjrngiyoueeunugsrhvcdgtzyvtggbhpocrbzmzithmddknjwbfxbpbiij2887736299497481439"): CreateProcess error=267, The directory name is invalid```

This is happening while initializing the workspace for git clone.

**Comment by MarkEWaite** [comment]
> `java.io.IOException: Cannot run program "git" (in directory "C:\Jenkins\workspace\Plugins_git-client-plugin_PR-523\target\tmp\tehdfmufzlrimdyxmzetepzxegbupzxewqtnlhwiwtcugemiizqsstahyyqufwpdjyolbkegykuhbwdgxmvezycngdpribpzvsiltzsdinfbokkwdnhsmfeeqcioeulxkqeagtjhqzuukqfglubmmcrbjhgifowrlbbxvbgdlqmuzjrngiyoueeunugsrhvcdgtzyvtggbhpocrbzmzithmddknjwbfxbpbiij2887736299497481439"): CreateProcess error=267, The directory name is invalid`
> 
> This is happening while initializing the workspace for git clone.

Probably indicates that the directory does not actually exist.  Likely the request to create the directory failed because the name of the directory was too long.  I believe Windows (and POSIX and Linux and BSD) limit the name of an individual file or directory to 255 characters.  At least that is what I observed on my Windows 10 computer.  I can create a directory with 255 characters in the name.  I can't create a directory with a 256 character name.


You may need to create two directories, one with a 200 character name, then a subdirectory with a name that is more than 60 characters.

**Comment by MarkEWaite** [comment]
@rishabhBudhouliya thanks again for this pull request.  In my CI infrastructure I had a test failing on windows due to not having long paths enabled.  I look forward to this extension so that the [submodule test](https://github.com/MarkEWaite/git-client-plugin/commits/enable-long-paths-in-submodule-test) in GitClient can use the extension to avoid the problem.

**Comment by rishabhBudhouliya** [comment]
@MarkEWaite thanks for the suggestions. While I took your advice to create two directories in (200 char + 60 char) format, I am still facing issues. 

- One of the mistakes I did earlier was to place the launchcommand for longpath **after** git init. I have corrected that mistake. 
- Since I don't have a Windows platform, would it be principally okay for me to commit on this PR and use the Jenkins CI to debug issues? 
- Or should I create my own Windows instance and rather work on that?

**Comment by rishabhBudhouliya** [comment]
This time I divided the directory structure, still I receive this issue: ```java.io.IOException: Cannot run program "git" (in directory "C:\Jenkins\workspace\Plugins_git-client-plugin_PR-523\target\tmp\dumxeaybjajaemmdzlbxgmhubedmkiwcpnbaugxafopmnmxbgudwfthfshdpzqebfqgeunsmndnteonxwdvgjtpugpvsvmxlabjztireqddsfurwsbjkraziyijmcdmlzggns2760806142667394690\gaizndyzlpowncyltbvvipxwdmmkqxrzamanfpgmpdpgbfflimraediaicyhbpvkorobigjgipvgbrpd"): CreateProcess error=267, The directory name is invalid```
